subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java

```diff
@@ -22,74 +22,215 @@
+    ) {
+        this.outputChangeListener = outputChangeListener;
+        this.outputsCleanerSupplier = outputsCleanerSupplier;
         this.delegate = delegate;
     }
 
     @Override
-    public CachingResult execute(UnitOfWork work, C context) {
-        ImmutableSortedMap<String, FileSystemSnapshot> previousOutputFiles = context.getPreviousExecutionState()
+    public CachingResult execute(UnitOfWork work, PreviousExecutionContext context) {
+        ImmutableSortedMap<String, CurrentFileCollectionFingerprint> knownFileFingerprints = context.getInputFileProperties();
+        ImmutableSortedMap<String, ValueSnapshot> knownValueSnapshots = context.getInputProperties();
+        InputFingerprinter.Result newInputs = fingerprintPrimaryInputs(work, context, knownFileFingerprints, knownValueSnapshots);
+
+        ImmutableSortedMap<String, CurrentFileCollectionFingerprint> sourceFileProperties = newInputs.getFileFingerprints();
+        if (!sourceFileProperties.isEmpty()) {
+            if (sourceFileProperties.values().stream()
+                .allMatch(CurrentFileCollectionFingerprint::isEmpty)
+            ) {
+                return skipExecutionWithEmptySources(work, context);
+            } else {
+                return executeWithNoEmptySources(work, context, newInputs.getAllFileFingerprints());
+            }
+        } else {
+            return executeWithNoEmptySources(work, context);
+        }
+    }
+
+    private InputFingerprinter.Result fingerprintPrimaryInputs(UnitOfWork work, PreviousExecutionContext context, ImmutableSortedMap<String, CurrentFileCollectionFingerprint> knownFileFingerprints, ImmutableSortedMap<String, ValueSnapshot> knownValueSnapshots) {
+        return work.getInputFingerprinter().fingerprintInputProperties(
+            context.getPreviousExecutionState()
+                .map(PreviousExecutionState::getInputProperties)
+                .orElse(ImmutableSortedMap.of()),
+            context.getPreviousExecutionState()
+                .map(PreviousExecutionState::getInputFileProperties)
+                .orElse(ImmutableSortedMap.of()),
+            knownValueSnapshots,
+            knownFileFingerprints,
+            visitor -> work.visitRegularInputs(new InputFingerprinter.InputVisitor() {
+                @Override
+                public void visitInputFileProperty(String propertyName, InputFingerprinter.InputPropertyType type, InputFingerprinter.FileValueSupplier value) {
+                    if (type == InputFingerprinter.InputPropertyType.PRIMARY) {
+                        visitor.visitInputFileProperty(propertyName, type, value);
+                    }
+                }
+            }));
+    }
+
+    @Nonnull
+    private CachingResult skipExecutionWithEmptySources(UnitOfWork work, PreviousExecutionContext context) {
+        context.getHistory()
+            .ifPresent(history -> history.remove(context.getIdentity().getUniqueId()));
```

Comment:
This should not be necessary anymore.

Reply:
Yeah. I'll rebase on your changes as soon as they hit `master`.